### PR TITLE
add the ability to retrieve the fieldnames an aggregation creates from the builder

### PR DESCRIPTION
**Pull request by hendrikmuhs**
add a getOutputFieldNames method to aggregationbuilder to retrieve the names
of the fields the aggregation produces

@polyfractal @nik9000 @not-napoleon this is a follow up from the discussion about adding the ability to retrieve the output field names of an aggregation from a builder instance.

I mainly added:

```
    /** Return the field names this aggregation creates*/
    public abstract Iterable<String> getOutputFieldNames();
```

and a set of common ones in `AbstractAggregationBuilder`.

The PR is not complete, I took some shortcuts and did not add any tests. I want to get the design 1st before spending to much resources on it.

Open discussion points (beside the interface):

 - how to handle the unknown case, e.g. `scripted_metric`, `top_hits`, ... Only report the top level key: `value`/`hits`?
 - sub-aggs are ignored in this change, I think that makes sense, does it?
 - key_as_string are ignored because they just contain a normalized value of another field
 - `extended_stats` has nested outputs, should it return e.g. `std_deviation_bounds.upper`?

Again, no need to review it in detail. Let's discuss the interfaces and some details. I will fix the rest once we nailed it down.

### CODE REVIEW

**Comment by not-napoleon** [comment]
It's unclear to me what the "right" set of field names is for a given aggregation.  For example, I would have expected `Filter` and `Filters` to have more or less the same output fields, but here we are saying `Filter` outputs doc counts only, while `Filters` outputs the names of the filters (and not doc count?). 

My biggest worry is that, as someone writing new aggregations, I would feel hard pressed to know what to return from `getOutputFieldNames()`.  Folks who aren't familiar with Transforms need an unambiguous way to correctly implement this function.  In a perfect world, we would define it in such a way that we could slap a test right on `AggregatorTestCase` to automatically verify the result of `getOutputFieldNames`, the way we are automatically checking the supported types there.

A second, smaller, concern - do we need to put `getOutputFieldNames` on `AggregatorBase` or would it be better to have an interface that only some builders implement?  That way, we could get rid of the `UNKNOWN` place holder, and just not implement the method for aggregations which can't know their output fields in advance.

If we can nail down how to explain to writers of new aggregations what the correct way to implement this new method is, I think this will be okay.  We've also talked (e.g. with the clients team) about having a formal API spec; I'm wondering if that would be an alternative solution to this issue.

**Comment by nik9000** [comment]
> A second, smaller, concern - do we need to put `getOutputFieldNames` on `AggregatorBase` or would it be better to have an interface that only some builders implement? That way, we could get rid of the `UNKNOWN` place holder, and just not implement the method for aggregations which can't know their output fields in advance.

I'm not a big fan of `instancof` to check to see if this is an aggregation we support. Returning `null` or `Optional.empty` or something would be ok.

I think this has to be closer to the builder than to `AggregatorBase` so we can do things with it on the coordinating node.

**Comment by not-napoleon** [comment]
Yeah, I hear you about the `instanceof` checks.  `Optional.empty` might be the way to go, I definitely like that better than just returning an empty collection.

**Comment by hendrikmuhs** [comment]
> It's unclear to me what the "right" set of field names is for a given aggregation

Yes, that's exactly my worry. It's very hard to define this. See the question in the start post.

> For example, I would have expected Filter and Filters to have more or less the same output fields, but here we are saying Filter outputs doc counts only, while Filters outputs the names of the filters (and not doc count?)

Filter only returns a doc_count, everything else comes from a sub-aggregation. I think this is correct, assuming we do not include the sub-agg in the output. Filters returns a field for every filter, however, should it return `filterA`, `filterB` or `filterA.doc_count`, `filterB.doc_count`, etc.

> Folks who aren't familiar with Transforms need an unambiguous way to correctly implement this function.

That's the wrong thinking. The change is motivated by transform, however this should not be transform method. If we fail to define the output, we failed. When we discussed this feature upfront we agreed, that this would also be helpful for other use cases (maybe EQL, PromQL, etc.). If we only aim for fixing transform, I have a work around for fixing `top_metric`.

> do we need to put getOutputFieldNames on AggregatorBase or would it be better to have an interface that only some builders implement

I actually only need it for `top_metrics`, however my impression from our discussion was, that we want a general solution. A partial implementation has risks and might yield surprises. Although for me, this would be easier, I think it's cleaner if we define it properly and implement a generic solution.

> That way, we could get rid of the UNKNOWN place holder

As said in the description, I took shortcuts, expect that this gets fixed. The PR is just a draft, going through the aggs and trying to implement as much as I could quickly, helped to see the problems. We should not have any unknown's in the end. But I am unsure how to solve `top_hits` and `scripted_metric`.

> Optional.empty might be the way to go,

Sounds good, looks like we need this for some cases, at least I do not have a solution for everything.

I wonder if getting the names is not enough. Maybe this functionality should return a best effort schema, instead of just names. E.g. a `doc_count` is a `long`(integer type), a key is a `string` etc. This can be abstract: whole number, floating point number, string, date, geo_tile, .... `UNKNOWN` would be `object` (we return something, we do not know what).

So instead of returning a list of strings, this interface could return a map, with the keys being the field names and the values being the expected output type. Now it gets interesting, there are aggregations which return whatever the source uses, e.g. a `max` returns a value based on whatever the source is. So the type for `max` would be an abstract type `source`.

Transform uses something very similar
https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/TransformAggregations.java#L95

This defines how to map the output of specific aggs to a data type.

**Comment by not-napoleon** [comment]
> I wonder if getting the names is not enough. Maybe this functionality should return a best effort schema, instead of just names. E.g. a doc_count is a long(integer type), a key is a string etc. This can be abstract: whole number, floating point number, string, date, geo_tile, .... UNKNOWN would be object (we return something, we do not know what).

I don't think we can make even reasonable guesses about the return types before resolving the values source types, so depending on what point in the build process you need this information that might not be available yet.   Even something as simple as string vs number, `Terms` agg has a mode where keys are numbers and we can't know that until we have the values source type.

**Comment by nik9000** [comment]
> I don't think we can make even reasonable guesses about the return types before resolving the values source types, so depending on what point in the build process you need this information that might not be available yet. Even something as simple as string vs number, `Terms` agg has a mode where keys are numbers and we can't know that until we have the values source type.

:+1: We can't be sure of the return type of all options.

If you need the type information too you really should wait until you've got the result I think. I don't think we could use *that* to validate pipeline aggs up front though.

Speaking of pipelines, @not-napoleon asked me to comment on the pipeline agg validation use case. I believe we'd talked about wanting to do "preflight" checks on the `bucketPaths` of the pipelines to make sure they actually refer to real things. Without the type information we can't be sure that they refer to things that the pipeline can consume, but we *can* check that the refer to real stuff.

**Comment by hendrikmuhs** [comment]
> I don't think we can make even reasonable guesses about the return types before resolving the values source types, so depending on what point in the build process you need this information that might not be available yet. Even something as simple as string vs number, `Terms` agg has a mode where keys are numbers and we can't know that until we have the values source type.

This got pulled out of context. Please read again what I wrote 2 sentences later, I already gave an answer for this:

> Now it gets interesting, there are aggregations which return whatever the source uses, e.g. a max returns a value based on whatever the source is. So the type for max would be an abstract type source.

Actually `terms` is not a good example: The output of terms is doc_count and buckets. The buckets are only of interest if there is a sub-aggregation. Nevertheless, it is correct that we can't deduct some types at build time, but only at runtime or using additional information like the mappings of the indexes to search on.

It remains "best effort", I don't think we can avoid placeholders (with and without output types).

To summarize the "best effort schema" again, the type information it could return:

- abstract semi-concrete type (string, count(non-floating point), number(floating point))
- same-as-source (output type == input type)
- object aka unknown

How should we proceed?

> I don't think we could use that to validate pipeline aggs up front though.

This sounds like your use case has vanished. Are there any other use cases you have in mind?

I still like the idea of the schema output. But: There is only 1 case it fixes which has a workaround. I have no problems with putting the idea on hold and implement the workaround.

**Comment by not-napoleon** [comment]
Summary of slack discussion:

- We think this is a useful feature that will help with validating pipeline bucket paths as well as transforms
- Types are a much harder problem than field names, and are somewhat less useful for the pipeline case.  For a v1, let's just do field names
- Proposed prototype:
```
/** Return the field names this aggregation creates*/
    public Optional<Set<String>> getOutputFieldNames() {
        return Optional.empty();
    }
```
- We like the idea of defaulting to returning `Optional.empty()` for a first pass.

**Comment by hendrikmuhs** [comment]
closing this in favor of #65139, I copied the summary of the last comment over to the new draft.
